File: simulation/combat/aura.go
package combat

import "github.com/Emyrk/chronicle/simulation/gamedata"

// Aura represents an active buff/debuff on a unit.
type Aura struct {
	SpellID       int32
	CasterLevel   int32
	Effects       [3]AuraEffect
	MaxDurationMs int32
	RemainingMs   int32
	StackCount    int32
	MaxStacks     int32
	ProcCharges   int32
	Permanent     bool
}

// AuraEffect represents a single effect within an aura.
type AuraEffect struct {
	Type       int32   // AuraType constant
	Amount     float64 // snapshotted value (damage per tick, stat bonus, etc.)
	MiscValue  int32   // school mask, stat type, etc.
	PeriodicMs int32   // tick interval (0 = non-periodic)
	NextTickMs int32   // ms until next tick
	Active     bool
}

// AuraTickEvent describes a periodic tick that occurred.
type AuraTickEvent struct {
	SpellID   int32
	EffectIdx int
	Amount    float64
	AuraType  int32
}

// AuraTracker manages active auras on a unit.
type AuraTracker struct {
	auras []*Aura
}

// NewAuraTracker creates an empty tracker.
func NewAuraTracker() *AuraTracker {
	return &AuraTracker{}
}

// AddAura adds or refreshes an aura. If an aura with the same SpellID
// already exists, it refreshes duration and adds a stack (up to MaxStacks).
func (t *AuraTracker) AddAura(a *Aura) {
	for _, existing := range t.auras {
		if existing.SpellID == a.SpellID {
			// Refresh duration
			existing.RemainingMs = a.MaxDurationMs
			if existing.MaxStacks > 0 && existing.StackCount < existing.MaxStacks {
				existing.StackCount++
			}
			// Update snapshotted amounts
			for i := range a.Effects {
				if a.Effects[i].Active {
					existing.Effects[i] = a.Effects[i]
				}
			}
			return
		}
	}
	if a.StackCount == 0 {
		a.StackCount = 1
	}
	t.auras = append(t.auras, a)
}

// RemoveAura removes the first aura matching spellID.
func (t *AuraTracker) RemoveAura(spellID int32) {
	for i, a := range t.auras {
		if a.SpellID == spellID {
			t.auras = append(t.auras[:i], t.auras[i+1:]...)
			return
		}
	}
}

// GetAura returns the aura for a spell, or nil.
func (t *AuraTracker) GetAura(spellID int32) *Aura {
	for _, a := range t.auras {
		if a.SpellID == spellID {
			return a
		}
	}
	return nil
}

// HasAura returns true if an aura with the given spellID is active.
func (t *AuraTracker) HasAura(spellID int32) bool {
	return t.GetAura(spellID) != nil
}

// ActiveAuras returns all active auras.
func (t *AuraTracker) ActiveAuras() []*Aura {
	return t.auras
}

// GetTotalModifier sums flat modifiers of a given aura type matching the school mask.
func (t *AuraTracker) GetTotalModifier(auraType int32, schoolMask int32) float64 {
	var total float64
	for _, a := range t.auras {
		for _, eff := range a.Effects {
			if !eff.Active || eff.Type != auraType {
				continue
			}
			if schoolMask == 0 || eff.MiscValue == 0 || (eff.MiscValue&schoolMask) != 0 {
				total += eff.Amount * float64(a.StackCount)
			}
		}
	}
	return total
}

// GetTotalModifierPercent returns the multiplicative product of percent mods
// of a given aura type matching the school mask.
func (t *AuraTracker) GetTotalModifierPercent(auraType int32, schoolMask int32) float64 {
	product := 1.0
	for _, a := range t.auras {
		for _, eff := range a.Effects {
			if !eff.Active || eff.Type != auraType {
				continue
			}
			if schoolMask == 0 || eff.MiscValue == 0 || (eff.MiscValue&schoolMask) != 0 {
				for s := int32(0); s < a.StackCount; s++ {
					product *= 1.0 + eff.Amount/100.0
				}
			}
		}
	}
	return product
}

// TickAuras advances all periodic effects by deltaMs, returning tick events.
func (t *AuraTracker) TickAuras(deltaMs int32) []AuraTickEvent {
	var events []AuraTickEvent
	for _, a := range t.auras {
		for i := range a.Effects {
			eff := &a.Effects[i]
			if !eff.Active || eff.PeriodicMs <= 0 {
				continue
			}
			eff.NextTickMs -= deltaMs
			for eff.NextTickMs <= 0 {
				events = append(events, AuraTickEvent{
					SpellID:   a.SpellID,
					EffectIdx: i,
					Amount:    eff.Amount * float64(a.StackCount),
					AuraType:  eff.Type,
				})
				eff.NextTickMs += eff.PeriodicMs
			}
		}
	}
	return events
}

// ExpireAuras decrements remaining duration by deltaMs and removes expired auras.
// Returns the spell IDs of expired auras.
func (t *AuraTracker) ExpireAuras(deltaMs int32) []int32 {
	var expired []int32
	remaining := t.auras[:0]
	for _, a := range t.auras {
		if a.Permanent {
			remaining = append(remaining, a)
			continue
		}
		a.RemainingMs -= deltaMs
		if a.RemainingMs <= 0 {
			expired = append(expired, a.SpellID)
		} else {
			remaining = append(remaining, a)
		}
	}
	t.auras = remaining
	return expired
}

// CreateAuraFromSpell creates an Aura from spell data, snapshotting periodic damage.
func CreateAuraFromSpell(spell *gamedata.SpellData, casterLevel int32, spellPower int32) *Aura {
	a := &Aura{
		SpellID:       spell.ID,
		CasterLevel:   casterLevel,
		MaxDurationMs: spell.DurationMs,
		RemainingMs:   spell.DurationMs,
		StackCount:    1,
	}
	for i := range spell.Effects {
		eff := &spell.Effects[i]
		if eff.Type == gamedata.SpellEffectNone {
			continue
		}
		if eff.Type == gamedata.SpellEffectApplyAura {
			ae := AuraEffect{
				Type:       eff.AuraType,
				Amount:     float64(eff.BasePoints + 1), // +1 for base dice
				MiscValue:  eff.MiscValue,
				PeriodicMs: eff.AuraPeriodMs,
				Active:     true,
			}
			if eff.AuraPeriodMs > 0 {
				ae.NextTickMs = eff.AuraPeriodMs
			}
			// Snapshot periodic damage with spell power
			if eff.AuraType == gamedata.AuraPeriodicDamage {
				coeff := float64(eff.BonusCoefficient)
				if coeff < 0 {
					// Default DoT coefficient
					numTicks := int32(0)
					if eff.AuraPeriodMs > 0 && spell.DurationMs > 0 {
						numTicks = spell.DurationMs / eff.AuraPeriodMs
					}
					coeff = DefaultSpellCoefficient(spell.CastTimeMs, spell.DurationMs, true, false, numTicks)
				}
				ae.Amount += float64(spellPower) * coeff * LevelPenalty(spell.SpellLevel)
			}
			a.Effects[i] = ae
		}
	}
	return a
}

File: frontend/chronicle/src/sim/aura.ts
/**
 * Aura tracking system. Ported from simulation/combat/aura.go.
 */

import { type SpellData, SpellEffectNone, SpellEffectApplyAura, AuraPeriodicDamage } from "./types";
import { defaultSpellCoefficient, levelPenalty } from "./formulas";

export interface AuraEffect {
  type: number;
  amount: number;
  miscValue: number;
  periodicMs: number;
  nextTickMs: number;
  active: boolean;
}

export interface Aura {
  spellID: number;
  casterLevel: number;
  effects: [AuraEffect, AuraEffect, AuraEffect];
  maxDurationMs: number;
  remainingMs: number;
  stackCount: number;
  maxStacks: number;
  procCharges: number;
  permanent: boolean;
}

export interface AuraTickEvent {
  spellID: number;
  effectIdx: number;
  amount: number;
  auraType: number;
}

function emptyEffect(): AuraEffect {
  return { type: 0, amount: 0, miscValue: 0, periodicMs: 0, nextTickMs: 0, active: false };
}

export class AuraTracker {
  private auras: Aura[] = [];

  addAura(a: Aura): void {
    for (const existing of this.auras) {
      if (existing.spellID === a.spellID) {
        existing.remainingMs = a.maxDurationMs;
        if (existing.maxStacks > 0 && existing.stackCount < existing.maxStacks) {
          existing.stackCount++;
        }
        for (let i = 0; i < 3; i++) {
          if (a.effects[i].active) existing.effects[i] = { ...a.effects[i] };
        }
        return;
      }
    }
    if (a.stackCount === 0) a.stackCount = 1;
    this.auras.push(a);
  }

  removeAura(spellID: number): void {
    const idx = this.auras.findIndex((a) => a.spellID === spellID);
    if (idx >= 0) this.auras.splice(idx, 1);
  }

  getAura(spellID: number): Aura | undefined {
    return this.auras.find((a) => a.spellID === spellID);
  }

  hasAura(spellID: number): boolean {
    return this.getAura(spellID) !== undefined;
  }

  activeAuras(): Aura[] {
    return this.auras;
  }

  getTotalModifier(auraType: number, schoolMask: number): number {
    let total = 0;
    for (const a of this.auras) {
      for (const eff of a.effects) {
        if (!eff.active || eff.type !== auraType) continue;
        if (schoolMask === 0 || eff.miscValue === 0 || (eff.miscValue & schoolMask) !== 0) {
          total += eff.amount * a.stackCount;
        }
      }
    }
    return total;
  }

  getTotalModifierPercent(auraType: number, schoolMask: number): number {
    let product = 1.0;
    for (const a of this.auras) {
      for (const eff of a.effects) {
        if (!eff.active || eff.type !== auraType) continue;
        if (schoolMask === 0 || eff.miscValue === 0 || (eff.miscValue & schoolMask) !== 0) {
          for (let s = 0; s < a.stackCount; s++) {
            product *= 1.0 + eff.amount / 100.0;
          }
        }
      }
    }
    return product;
  }

  tickAuras(deltaMs: number): AuraTickEvent[] {
    const events: AuraTickEvent[] = [];
    for (const a of this.auras) {
      for (let i = 0; i < 3; i++) {
        const eff = a.effects[i];
        if (!eff.active || eff.periodicMs <= 0) continue;
        eff.nextTickMs -= deltaMs;
        while (eff.nextTickMs <= 0) {
          events.push({
            spellID: a.spellID,
            effectIdx: i,
            amount: eff.amount * a.stackCount,
            auraType: eff.type,
          });
          eff.nextTickMs += eff.periodicMs;
        }
      }
    }
    return events;
  }

  expireAuras(deltaMs: number): number[] {
    const expired: number[] = [];
    this.auras = this.auras.filter((a) => {
      if (a.permanent) return true;
      a.remainingMs -= deltaMs;
      if (a.remainingMs <= 0) {
        expired.push(a.spellID);
        return false;
      }
      return true;
    });
    return expired;
  }
}

/** Create an Aura from spell data, snapshotting periodic damage with spell power. */
export function createAuraFromSpell(spell: SpellData, casterLevel: number, spellPower: number): Aura {
  const a: Aura = {
    spellID: spell.id,
    casterLevel,
    effects: [emptyEffect(), emptyEffect(), emptyEffect()],
    maxDurationMs: spell.durationMs,
    remainingMs: spell.durationMs,
    stackCount: 1,
    maxStacks: 0,
    procCharges: 0,
    permanent: false,
  };

  for (let i = 0; i < 3; i++) {
    const eff = spell.effects[i];
    if (eff.type === SpellEffectNone) continue;
    if (eff.type === SpellEffectApplyAura) {
      const ae: AuraEffect = {
        type: eff.auraType,
        amount: eff.basePoints + 1,
        miscValue: eff.miscValue,
        periodicMs: eff.auraPeriodMs,
        nextTickMs: eff.auraPeriodMs > 0 ? eff.auraPeriodMs : 0,
        active: true,
      };
      if (eff.auraType === AuraPeriodicDamage) {
        let coeff = eff.bonusCoefficient;
        if (coeff < 0) {
          let numTicks = 0;
          if (eff.auraPeriodMs > 0 && spell.durationMs > 0) {
            numTicks = Math.floor(spell.durationMs / eff.auraPeriodMs);
          }
          coeff = defaultSpellCoefficient(spell.castTimeMs, spell.durationMs, true, false, numTicks);
        }
        ae.amount += spellPower * coeff * levelPenalty(spell.spellLevel);
      }
      a.effects[i] = ae;
    }
  }

  return a;
}
